Add support for succinct roles (TAP 15)

MVrachev opened the pull request:
Related to the previous pr: https://github.com/theupdateframework/python-tuf/pull/1948 and issue #1909

**This pr includes 11 breaking changes:**

In `Delegations` class from `tuf/api/metadata.py` :
1. roles argument is made optional
2. unrecognized_fields argument becomes the 4-th rather than the 3-rd
as it used to be

In `Root` class from `tuf/api/metadata.py`:
1. The "role" and "key" arguments in "Root.add_key()" are in reverse
order - "key" becomes first and "role" second.
2. "Root.remove_key()" has been renamed to "Root.remove_key()"
3. The "role" and "keyid" arguments in "Root.revoke_key()" are in
reverse order - "keyid" becomes first and "role" second.

In `Targets` class from `tuf/api/metadata.py`:
1. The "role" and "key" arguments in "Targets.add_key()" are in reverse
order - "key" becomes first and "role" second
4. "Targets.remove_key()" has been renamed to "Targets.revoke_key()"
5. The "role" and "keyid" arguments in "Targets.revoke_key()" are in
reverse order - "keyid" becomes first and "role" second.
6.  In both methods "Targets.add_key()" and "Targets.revoke_key()" the
"role" argument becomes an optional with a default value of None

Those changes are made in an effort to make those two methods logical
for both cases when standard roles and succinct_roles are used.

**Description of the changes being introduced by the pull request**:

[TAP 15](https://github.com/theupdateframework/taps/blob/master/tap15.md) was created on June 23-rd 2020 and was last modified on July 6-th 2020.
Since then the TAP has been put to `draft` status meaning it needs a prototype implementation before it's accepted as a future specification change.
Given that this TAP underlines an efficient way of handling succinct hash bin delegations, meaning it would be really useful when `python-tuf` is integrated into `Warehouse`, it's logical that we should not only create a prototype but directly work to integrate it in `python-tuf`.

The outcome of the TAP 15 implementation is
1. Support reading, updating, and saving target metadata files that are using `succinct_roles`.
3. Support in ngclient for calculating the delegated role name on a given target path.
4. An exemplary document inside `examples/` showcasing how `succinct_roles` can be utilized in practice.

This pr covers points 1 and 2.
Point 3 will be done in another pr.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Add SuccinctRole class

Add SuccinctRoles class containing the information from the
succint_roles dict described in TAP 15.
This allows for easy mypy checks on the types, easy enforcement on
TAP 15 restrictions (as for example that "bit_length" must be between 1
and 32) and support for unrecognized fields inside succinct_roles
without much of a hassle.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
SuccinctRoles: add zero padding to bins

Add zero padding to bin names inside SuccinctRoles.
Zero padding ensures that the bin names always have the same length.

This characteristic is implied in the example given by TAP 15 where
the third bin is named "alice.hbd-03". For context read TAP 15:
https://github.com/theupdateframework/taps/blob/master/tap15.md

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add helper methods in SuccinctRoles

Add two helper methods in SuccinctRoles.
Those methods proved useful in the testing code, but I believe they have
a potential value for production code as well.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Clarifications and simplifications

Clarify explicitly that exactly one of "paths" and "path_hash_prefixes"
must be set inside DelegatedRole.
Also simplify the check for "paths" and "path_hash_prefixes".
Finally, add a test case inside the "test_metadata_serialization.py"
test file about wrong keyids type for "Role" serialization.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Test delegation tree with succinct_roles

Test traversing the delegation tree when there is a Targets using a
delegation with succinct roles.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I know the code is a lot, but a big chunk of it is testing.

Commit:
SuccinctRoles: add readme information

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1444,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
+        """
+        desired_prefix = self.name_prefix + "-"
+
+        if role_name.startswith(desired_prefix):
+            suffix = role_name[len(desired_prefix) :]
+            try:
+                # bin name suffixes are in hex format.
+                num = int(suffix, 16)
+                if (
+                    0 <= num < self.number_of_bins
+                    and len(suffix) == self.suffix_len
+                ):
+                    return True
+            except ValueError:
+                # suffix is not a number
+                return False
+
+        return False
```
this would likely improve from early returns, something like this:

```suggestion
        if not role_name.startswith(desired_prefix):
            return False

        suffix = role_name[len(desired_prefix) :]
        if len(suffix) != self.suffix_len:
            return False

        try:
            # make sure suffix is hex value
            num = int(suffix, 16)
        except ValueError:
            return False

        return num >= 0 and num < self.number_of_bins
```

MVrachev replied:
Agree and it looks more readable.
I heavily focused on what is necessary for the `role_name` to be a bin and ended up with long checks.
BTW: `0 <= num < self.number_of_bins` was suggested by `pylint` I think instead of `num >= 0 and num < self.number_of_bins`.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1457,18 +1611,24 @@ class Delegations:
-                    "Delegated roles cannot be empty or use top-level role names"
-                )
+        if sum(1 for v in [roles, succint_roles] if v is not None) != 1:
+            raise ValueError("Only one of roles and succinct_roles must be set")
```
This error is also raised when neither are set. Then the error message might be surprising.

jku commented on tuf/api/metadata.py:
```diff
@@ -1643,25 +1665,78 @@ def from_dict(cls, delegations_dict: Dict[str, Any]) -> "Delegations":
+            target_filepath: URL path to a target file, relative to a base
+                targets URL.
+        """
+        if isinstance(self.roles, dict):
```
There are multiple isinstance() calls that look a little fishy to me -- they imply there are cases where `self.roles` type is something other than Optional[Dict] which is not true. I think`self.roles is not None` is clearer and also matches the elif a few lines after it.

this comment likely applies most isinstance() calls in the actual code -- tests can do whatever they want.

MVrachev replied:
That's a good catch.
I thought that I need to verify that `self.roles` is actually a `Dict` in order for `mypy` to understands why I am using dictionary methods, but you are right that `self.roles` is defined clearly as `Optional[Dict]` and it shouldn't be something else rather than `None` or `Dict`.

jku commented on tuf/api/metadata.py:
```diff
@@ -1643,25 +1665,78 @@ def from_dict(cls, delegations_dict: Dict[str, Any]) -> "Delegations":
+            res_dict["succinct_roles"] = self.succinct_roles.to_dict()
+        return res_dict
+
+    def find_delegation(self, target_filepath: str) -> Optional[str]:
```
is find_delegation() needed somewhere?

jku commented on tuf/api/metadata.py:
```diff
@@ -1603,18 +1611,24 @@ class Delegations:
         keys: Dict[str, Key],
-        roles: Dict[str, DelegatedRole],
+        roles: Optional[Dict[str, DelegatedRole]] = None,
+        succint_roles: Optional[SuccinctRoles] = None,
```
typo succinct

jku commented on tuf/api/metadata.py:
```diff
@@ -1603,18 +1611,24 @@ class Delegations:
-        roles: Dict[str, DelegatedRole],
+        roles: Optional[Dict[str, DelegatedRole]] = None,
+        succint_roles: Optional[SuccinctRoles] = None,
         unrecognized_fields: Optional[Dict[str, Any]] = None,
```
This is an API change -- it's likely an API change we want to do but this sort of things really should be pointed out in the commit message and PR cover letters.

The two changes here are moving unrecognised fields in constructor argument list and more importantly making the roles attribute optional -- this is mostly relevant to code that uses `Delegations.roles` as it now has to check for None (As far as I can tell you've done a great job with handling all of that code within our codebase but any tuf using code, e.g. repository-editor-for-tuf, now needs to add None-checks)

MVrachev replied:
Added a sentence in the commit message about those two API changes.

jku commented on tuf/api/metadata.py:
```diff
@@ -1593,6 +1599,8 @@ class Delegations:
             role. The roles order also defines the order that role delegations
             are considered during target searches.
+        succinct_roles: Contains succinct information about hash bin
+            delegations. For more information see: https://github.com/theupdateframework/taps/blob/master/tap15.md
```
the fact that exactly one of [roles, succinct_roles] is not-None should be documented

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1497,25 +1665,78 @@ def from_dict(cls, delegations_dict: Dict[str, Any]) -> "Delegations":
+            res_dict["succinct_roles"] = self.succinct_roles.to_dict()
+        return res_dict
+
+    def find_delegation(self, target_filepath: str) -> Optional[str]:
```
Can we remove this API function, given that it's nowhere used?

MVrachev replied:
Currently, It's not used in the code, but my reasoning was that I thought that having a method that gives you the correct delegation now when we support succinct roles and standard roles seemed a must-have.
Now, when I think about it we have `get_all_delegations` which does actually the same thing, but instead of returning just one role, it returns all roles responsible for that `target_filepath` which could be many in the case of standard roles.

You are right, we have a replacement for this method, and given it's not currently used it's better to remove it.
Again, we can always add it later if the need arises.

jku commented on tuf/api/metadata.py:
```diff
@@ -2012,6 +2012,31 @@ def add_key(self, role: str, key: Key) -> None:
 
         self.delegations.keys[key.keyid] = key
 
+    def add_key_in_succinct_roles(self, key: Key) -> None:
```
I'm not a big fan of this one: it adds 50 lines of code (where 2 lines is new functionality) and two new public methods ... and reality is that the old methods could work just fine, the API just looks a bit odd with the unused rolename.

The PR as a whole is already an API change so let's consider modifying the API here as well, and let's also consider the option of just documenting that rolename can be an unused argument when calling `add_key()`.

jku replied:
You say in the comment about this 

> clarify what the role name for succinct_roles should be (but this
> should happen in the TAP 15 itself and will lead to unnecessary
> complications)

I don't agree. `add_key()` is our API and unrelated to the TAP, and the problem is completely in our API decisions. We can choose how it works.

jku commented on tests/repository_simulator.py:
```diff
@@ -377,6 +378,44 @@ def add_delegation(
+                {}, roles=None, succinct_roles=succinct_roles
+            )
+        else:
+            delegator.delegations.succinct_roles = succinct_roles
```
style comment: I would make this shorter and easier to understand, like:

        key, signer = self.create_key()
        succinct_roles = SuccinctRoles([], 1, bit_length, name_prefix)
        delegator.delegations = Delegations({}, None, succinct_roles)
 
 it's a testing helper class so I don't think handling every weird case is possible or that useful. But less code is useful for readers

jku commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1444,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
+            **base_role_dict,
+        }
+
+    def get_bin_name(self, bin_number: int = 0) -> str:
```
is there a reason for this to be a public method?

jku replied:
also the default value seems to not be useful or used.

MVrachev replied:
I was wondering whether I should leave this method as I thought it could be used by users of TUF.
But also if we see the need for it we can always add it.
Will remove it.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1444,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
 
+class SuccinctRoles(Role):
+    """A container with succinct information about hash bin delegation for all
+    possible target paths. It's can be used inside ``Delegations``.
```
```suggestion
    possible target paths. It can be used inside ``Delegations``.
```

lukpueh replied:
Minor doc style nit: The summary of a class/function should fit on one line. Same comment applies elsewhere in this PR.

jku reviewed (commented):
Overall this looks like a good direction I think and code does not seem to have issues (bugs). I have left some individual code comments (no bugs or anything, just opinions). These are the concerns:
* this adds a large amount of new API, 10 new methods or something like that: let's focus on whether all of that is actually used and needed, whether it needs to be public API and whether there are other options to adding the API (ie the `add_key()` case). I've left some comments 
* this is also an API change: I think it's one we want to do but let's take changes seriously

I've not checked if the test coverage is "complete": let's figure out the API size first.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1444,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
 
 
+class SuccinctRoles(Role):
+    """A container with succinct information about hash bin delegation for all
+    possible target paths. It's can be used inside ``Delegations``.
+
+    Succinct roles describes delegations for all possible target paths
+    to a variable number of bins, the names of which are generated based on the
+    two class attributes.
+
+    Args:
+        keyids: Signing key identifiers.
+        threshold: Number of keys required to sign succinct roles metadata.
+        bit_length: Number of bits between 1 and 32 used to separate the
+            different bins.
+        name_prefix: Prefix of all bin names.
+        unrecognized_fields: Dictionary of all attributes that are not managed
+            by TUF Metadata API.
```
I still think the entire class docstring requires a lot of prior knowledge about succinct hash delegation, to a point where it isn't relevant anymore. 

Would do you think about the following suggestion. It is slightly wordier, but also provides more context from TAP 15:

```suggestion
    """Succinctly defines a hash bin delegation graph.

    A ``SuccinctRoles`` object describes a delegation graph that covers all targets,
    distributing them uniformly over the delegated roles (i.e. bins) in the graph.

    The total number of bins is 2 to the power of the passed ``bit_length``. And targets
    are assigned to bins by casting the left-most ``bit_length`` bits of the file path
    hash digest to int, using it as bin index between 0 and ``2**bit_length - 1``.

    Bin names are the concatenation of the passed ``name_prefix`` and a zero-padded hex
    representation of the bin index between separated by a hyphen.

    The same passed ``keyids`` and ``threshold`` is used for each bin, and each bin is
    'terminating'.

    See https://github.com/theupdateframework/taps/blob/master/tap15.md for details.

    Args:
        keyids: Signing key identifiers for any bin metadata.
        threshold: Number of keys required to sign any bin metadata.
        bit_length: Number of bits between 1 and 32.
        name_prefix: Prefix of all bin names.
        unrecognized_fields: Dictionary of all attributes that are not managed
            by TUF Metadata API.
```

lukpueh replied:
Feel free to ignore/edit as you please.

MVrachev replied:
Overall, I like the suggestion.
Most of the sentences I took as they and others I used with small modifications.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1444,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
+        """Return the name of bin with number ``bin_number``.
+
+        The ``bin_number`` default is 0 as that is used to uniquely identify
+        delegated roles containing succinct_roles from the rest.
```
This seems like an artifact from the first iteration of the PR. Do we still need it? If not, I suggest to remove the default.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1444,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
+        Args:
+            bin_number: Number of the bin that we want the name for.
+        """
+        # Add zero padding if necessary and cast to hex the suffix.
```
Word order seems off in this comment

jku commented:
Documenting some possible options WRT `add_key()` and `remove_key()`:
* as in PR: add new methods for succinct case (so users need to remember to call different Targets method depending on what the value of Targets.delegations.succinct_delegations is)
* just document that `role` argument to the `*_key()` functions is unused whenever succinct delegations are used and make the existing functions work in succinct case
* Modify `*_key()` API so that instead of `self, role: str, key: Key` the arguments are `self, key: Key, role: Optional[str] = None`

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1497,25 +1665,78 @@ def from_dict(cls, delegations_dict: Dict[str, Any]) -> "Delegations":
+        elif self.succinct_roles is not None:
+            # We consider all succinct_roles as terminating.
+            # For more information read TAP 15.
+            yield self.succinct_roles.find_bin(target_filepath), True
```
Is it worth adding `else: raise RuntimeError`?

MVrachev replied:
Well, I rely on the idea that the user has created the `Delegations` object and then not deliberately changed the object in such a way that `roles` and `succinct_roles` are both `None`.
If a user has used the constructor and is in a situation where `succinct_roles` and `roles` is None, then he/she will be notified.

In short I am not sure we need that here.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1773,13 +1994,47 @@ def add_key(self, role: str, key: Key) -> None:
+        """
+        if (
+            self.delegations is None
+            or self.delegations.roles is not None
```
IMO this method shouldn't care about `self.delegations.roles`.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1791,16 +2046,57 @@ def remove_key(self, role: str, keyid: str) -> None:
+        if self.delegations.succinct_roles is not None:
+            raise ValueError(
+                "Don't use remove_key when succinct_roles is in use"
+            )
```
IMO this method shouldn't care about `self.delegations.succinct_roles`.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1773,13 +1994,47 @@ def add_key(self, role: str, key: Key) -> None:
-            self.delegations.roles[role].keyids.append(key.keyid)
+
+        if self.delegations.succinct_roles is not None:
+            raise ValueError("Don't use add_key when succinct_roles is in use")
```
IMO this method shouldn't care about `self.delegations.succinct_roles`.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1791,16 +2046,57 @@ def remove_key(self, role: str, keyid: str) -> None:
+        """
+        if (
+            self.delegations is None
+            or self.delegations.roles is not None
```
IMO this method shouldn't care about `self.delegations.roles`.

lukpueh reviewed (changes requested):
This is really great, @MVrachev. I left a few comments inline, but nothing big a deal.

I also have a few more naming suggestions for `SuccinctRoles` methods, to consolidate with the rest of the API...
```
get_bin_name --> get_role(_for_index)
get_all_bin_names --> get_roles
find_bin --> get_role_for_target
is_bin --> is delegated_role
```
... and for `Delegations`:
```
get_all_delegations -> get_roles_for_target
```

What do you think?

MVrachev commented:
> get_bin_name --> get_role(_for_index)
> get_all_bin_names --> get_roles
> find_bin --> get_role_for_target
> is_bin --> is delegated_role

I understand where you come from and why those changes make sense for you.
I am worried about using the word `role` inside the `SuccinctRole` method names and especially the phrase `delegated role` because I don't want to confuse our users with `Delegations.roles` or the `DelegatedRole` class instances.
That's why I opt-in for a more concrete naming scheme including `bin` instead of role.
Also, after your docstring suggestion, it seems a lot clearer what a bin actually is.

After I update the pr with your docstring suggestions would you have a look again and tell me if you think those changes are still worth it?

Commit:
RepositorySimulator: support succinct_roles

Add support for Targets using delegation with succinct_roles.
For that purpose, we needed a method that can add succinct_roles
information with its all corresponding bins to the target metadata
and self.md_delegates attribute in RepositorySimulator.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Delegations: add support for succinct_roles

This commit contains 2 API changes in "Delegations" class from
tuf/api/metadata.py:
1. roles argment is made optional
2. unrecognized_fields argument becomes the 4-th rather than the 3-rd
as it used to be

In this commit, I add support for succinct_roles roles inside
Delegations class. This change is related to TAP 15 proposal.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Root and Targets key API changes

Here is the list of all breaking API changes:
1) The "role" and "key" arguments in "Root.add_key()" are in reverse
order - "key" becomes first and "role" second.
2) "Root.remove_key()" has been renamed to "Root.revoke_key()".
3) The "role" and "keyid" arguments in "Root.revoke_key()" are in
reverse order - "keyid" becomes first and "role" second.
4) The "role" and "key" arguments in "Targets.add_key()" are in reverse
order - "key" becomes first and "role" second.
5) "Targets.remove_key()" has been renamed to "Targets.revoke_key()".
6) The "role" and "keyid" arguments in "Targets.revoke_key()" are in
reverse order - "keyid" becomes first and "role" second.
7) In both methods "Targets.add_key()" and "Targets.revoke_key()" the
"role" argument becomes an optional with a default value of None.

Those changes are made in an effort to make those methods logical
for both cases when standard roles and succinct_roles are used.
The "Root" API change was done in order to preserve naming and argument
order consistency with "Targets" API.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have addressed all of your comments @jku and @lukpueh.
For `Targets.*_key()` methods I decided to change the order of the arguments and make `role` an optional.
One downside in this (besides it's a breaking change) is that those methods no longer have the same signature as `Root.*_key()` methods.

Also, I have mentioned in the commit messages about the breaking API changes and also updated the pr description with information about them.

The pr is smaller by a big margin and it's ready for another review.

lukpueh commented:
> > get_bin_name --> get_role(_for_index)
> > get_all_bin_names --> get_roles
> > find_bin --> get_role_for_target
> > is_bin --> is delegated_role
> 
> I understand where you come from and why those changes make sense for you. I am worried about using the word `role` inside the `SuccinctRole` method names and especially the phrase `delegated role` because I don't want to confuse our users with `Delegations.roles` or the `DelegatedRole` class instances. That's why I opt-in for a more concrete naming scheme including `bin` instead of role. Also, after your docstring suggestion, it seems a lot clearer what a bin actually is.
> 
> After I update the pr with your docstring suggestions would you have a look again and tell me if you think those changes are still worth it?

My idea was to point out exactly that connection, i.e. that a bin is a delegated targets role. But I won't insist. I'm fine with the names you picked as well.

What about my other suggestion `get_all_delegations -> get_roles_for_target`? The idea behind this one is that the function does not really return delegations but delegated roles (or actually their names, and their terminating property) 

I'm a bit pedantic about this, because roles and delegations are confused so often.

MVrachev commented:
> My idea was to point out exactly that connection, i.e. that a bin is a delegated targets role. But I won't insist. I'm fine with the names you picked as well.
> 
> What about my other suggestion `get_all_delegations -> get_roles_for_target`? The idea behind this one is that the function does not really return delegations but delegated roles (or actually their names, and their terminating property)
> 
> I'm a bit pedantic about this, because roles and delegations are confused so often.

Hmm... more and more when I think about it probably you have some point here.
I keep considering the `SuccinctRoles` instances as one delegated role which is not what is implied even from the class title... `SuccinctRoles` instance just represents a group of delegated roles instead of using a dictionary. Probably mentally it's easier to think `SuccinctRoles` is just one role as `keyids` and `threshold` is shared through all delegated roles represented by that instance.
Probably it will be useful in using the same `delegated role` phrase for `SuccinctRoles` roles as well instead of promoting a new `bin` term.

Will rename the methods as you have suggested here: 
> get_all_bin_names --> get_roles
> find_bin --> get_role_for_target
> is_bin --> is delegated_role

About `get_all_delegations -> get_roles_for_target`.
Yea, I missed it honestly as there were so many renames. I agree with that suggestion. Will do it.

MVrachev commented:
I followed @lukpueh advice and did the following renames:
- `get_all_bin_names` --> `get_roles`
- `find_bin` --> `get_role_for_target`
- `is_bin` --> `is delegated_role`
- `get_all_delegations` -> `get_roles_for_target`

EDIT: I forced pushed once more to fix the commit message mentioning the old methods name.

jku commented:
Looks good. The only thing that sticks out to me is the key API.

```
- def add_key(self, role: str, key: Key) -> None:
+ def add_key(self, key: Key, role: Optional[str] = None) -> None:
```

this looks ok to me -- it's easy to notice with static checks and we could even add a specific error message `if isinstance(role, Key):`...

```
- def remove_key(self, role: str, keyid: str) -> None:
+ def remove_key(self, keyid: str, role: Optional[str] = None) -> None:
```
... but I hadn't thought about this one all the way through: both argument types are str here. It really is the most annoying kind of API change, very easy to mess up.

MVrachev commented:
> Looks good. The only thing that sticks out to me is the key API.
> 
> ```
> - def add_key(self, role: str, key: Key) -> None:
> + def add_key(self, key: Key, role: Optional[str] = None) -> None:
> ```
> 
> this looks ok to me -- it's easy to notice with static checks and we could even add a specific error message `if isinstance(role, Key):`...

Maybe this could be useful. Preventing users from mistaking the arguments order.

> 
> ```
> - def remove_key(self, role: str, keyid: str) -> None:
> + def remove_key(self, keyid: str, role: Optional[str] = None) -> None:
> ```
> 
> ... but I hadn't thought about this one all the way through: both argument types are str here. It really is the most annoying kind of API change, very easy to mess up.

Yea... if only the types were different. I was thinking if we can use the insider knowledge we have about `keyid`.
For example that it will have a length longer than X number of symbols, but then what stops a user from having very long delegated role names? Nothing...

I hope we can make this API change as still there aren't so many users, especially in the 1.0.0 version of python-tuf and calls like `targets.add_key(None, key_obj)` and  ``targets.add_key(None, keyid)` are not really nice...

jku commented:
> I was thinking if we can use the insider knowledge we have about keyid.

I don't think we have any insider knowledge about it -- theoretically it's a hash but
* we've explicitly avoided using that in python-tuf as it seems like a mistake: it should be just a unique id 
* even if we know keyid is a hash, nothing prevents rolenames from also being hashes (in fact I could imagine it would be a good way to handle rolenames in some cases)

MVrachev commented:
@lukpueh any thoughts?

jku commented:
I think there's still several possibilities here that wouldn't require surprisingly breaking applications:
* keep current API (maybe allow None for rolename), just document it well enough. I think this is completely within reason. It's a wart in the API but that happens.
* `@overload` magic so that we could have both `remove_key(rolename, keyid)` and `remove_key(keyid)` -- the code will not be very pretty but should work.
* just rename the API: remove_key() stays as is for a few releases and we have a new method with different API -- the lovely linux naming tradition would give us `remove_key2()` :grimacing: 
* probably something else still?

lukpueh commented:
> I think there's still several possibilities here that wouldn't require surprisingly breaking applications:
> 
> * keep current API (maybe allow None for rolename), just document it well enough. I think this is completely within reason. It's a wart in the API but that happens.

This would be an okay solution.

> * `@overload` magic so that we could have both `remove_key(rolename, keyid)` and `remove_key(keyid)` -- the code will not be very pretty but should work.

I'm not sure about this one. It is definitely unidiomatic (~> surprising). And the code to do this would likely stain the otherwise elegant metadata module.

> * just rename the API: remove_key() stays as is for a few releases and we have a new method with different API -- the lovely linux naming tradition would give us `remove_key2()` 😬

I like this suggestion best, at least the first part. But I'd rather not have a `remove_key2` in our API. And I also don't think we need to support the old method for a few releases, if we make this release a breaking one.

So what about replacing `remove_key` right away with a name that also makes sense and that we like. Some ideas: `revoke_key`, `deauthorize`, `unauthorize`, *something with delegation in its name*, etc...

Some consistency considerations:
- Will the new name work as counterpart to `add_key` or do we need to rename that too?
- What about `Root`'s `{add, remove}_key` methods. Should we adopt a rename there too? And the argument order? This would make it a bigger cut, but API consistency seems important.

> * probably something else still?

Couldn't think of anything else.

jku commented:
Yes, agreed that providing the old deprecated method isn't terribly important if using the new method is basically a copy-paste fix

lukpueh commented on examples/repo_example/hashed_bin_delegation.py:
```diff
@@ -162,6 +162,7 @@ def find_hash_bin(path: str) -> str:
 #              10-17                       10 11 12 13 14 15 16 17
 #              ...                         ...
 #              f8-ff                       f8 f9 fa fb fc fd fe ff
+assert isinstance(roles["bins"].signed.delegations.roles, dict)
```
Why assert this here?

MVrachev replied:
Because of `mypy` I have to assert that `roles["bins"].signed.delegations.roles` is None as a few lines later I have an assignment. 
Will change the assert to verify that `roles["bins"].signed.delegations.roles is not None` as if it's not none it can only be a dict.

lukpueh commented on tests/test_api.py:
```diff
@@ -517,46 +519,89 @@ def test_targets_key_api(self) -> None:
+
+        # Remove delegated roles.
+        assert isinstance(targets.delegations, Delegations)
+        assert targets.delegations.roles is not None
```
Why `assert` and not `self.assert*`?

MVrachev replied:
`mypy` doesn't recognize `self.assert*` methods.
For example, if I replace `assert isinstance(targets.delegations, Delegations)` with `self.assertIsNotNone(targets.delegations)` then I multiple warnings like:
`error: Item "None" of "Optional[Delegations]" has no attribute "roles"  [union-attr]`

lukpueh commented on tests/test_api.py:
```diff
@@ -695,6 +740,38 @@ def test_is_delegated_role(self) -> None:
             self.assertTrue(role.is_delegated_path("a/path"))
 
+    def test_zero_padding_in_succinct_roles(self) -> None:
+        bit_lengths = [2, 6, 8, 10, 16, 32]
```
Any reason for these exact bit lengths? Why not test `range(1, 32)`?

MVrachev replied:
I wanted to choose a subset of key `bit_lenghts` in which the expected suffix length will be different if I want to test against `range(1, 32)` I will have to come up with a formula about the expected suffix length (which even if I do it won't be so readable and maintainable) or have a list of 32 `bit_lengths` and 32 expected suffix lengths which I am not sure it makes sense.

lukpueh commented on tests/test_api.py:
```diff
@@ -695,6 +740,38 @@ def test_is_delegated_role(self) -> None:
 
+    def test_zero_padding_in_succinct_roles(self) -> None:
+        bit_lengths = [2, 6, 8, 10, 16, 32]
+        expected_paddings = [1, 2, 2, 3, 4, 8]
```
Minor naming nit: These are not paddings.

MVrachev replied:
Agree, will rename it to `expected_suffix_lengths`.

lukpueh commented on tests/test_api.py:
```diff
@@ -695,6 +740,38 @@ def test_is_delegated_role(self) -> None:
+
+    def test_is_delegated_role_in_succinct_roles(self) -> None:
+        succinct_roles = SuccinctRoles([], 1, 5, "bin")
+        # delegated role name suffixes are in hex format.
```
"bin-20" and "bin-100" are valid hex numbers, so the comment feels a bit odd

MVrachev replied:
Okay, will move this comment before the `true_name_examples` and then it will make more sense.

lukpueh commented on tests/test_api.py:
```diff
@@ -695,6 +740,38 @@ def test_is_delegated_role(self) -> None:
+        bin_amount = 0
+        for i, role_name in enumerate(succinct_roles.get_roles()):
+            self.assertEqual(role_name, f"bin-{i}")
+            bin_amount = i
```
`bin_amount` does not seem very useful. :)

MVrachev replied:
Okay, will remove it.

lukpueh commented on tests/test_api.py:
```diff
@@ -695,6 +740,38 @@ def test_is_delegated_role(self) -> None:
+        succinct_roles = SuccinctRoles([], 1, 3, "bin")
+        bin_amount = 0
+        for i, role_name in enumerate(succinct_roles.get_roles()):
+            self.assertEqual(role_name, f"bin-{i}")
```
This assert would become more interesting for a bit_length > 3, which would implicitly also test the padding and you wouldn't need `test_zero_padding_in_succinct_roles`.

MVrachev replied:
Okay, will do it.

lukpueh commented on tests/test_metadata_eq_.py:
```diff
@@ -220,6 +221,19 @@ def test_delegated_role_eq_(self) -> None:
             # Restore the old value of the attribute.
             setattr(delegated_role_2, attr, getattr(delegated_role, attr))
 
+    def test_succinct_roles_eq(self) -> None:
```
Let's merge #2017 first and update this test accordingly, shall we?

lukpueh reviewed (changes requested):
Great tests, @MVrachev. I just reviewed:

```
docs/api/tuf.api.metadata.supporting.rst
docs/repository-library-design.md
examples/repo_example/hashed_bin_delegation.py
tests/test_api.py
tests/test_metadata_eq_.py
```

I'll review the rest tomorrow.

MVrachev commented:
I think `revoke_key` and the counterpart `add_key` sounds okay.

If we decide to do a rename to `remove_key` then does this mean that we are okay with its counterpart - `add_key` and its changes after I add a check from the type `if isinstance(role, Key)` as Jussi has suggested preventing mistakes?

lukpueh commented:
Any thoughts about also changing `Root.{add, remove}_key`?

lukpueh commented on tests/repository_simulator.py:
```diff
@@ -354,21 +355,60 @@ def add_delegation(
-            delegator.delegations = Delegations({}, {})
+            delegator.delegations = Delegations({}, roles={})
+
+        assert delegator.delegations.roles is not None
```
Is an `AssertionError` here better than a `TypeError` in the next line?

MVrachev replied:
I do this because of mypy. Otherwise, it gives a warning:
`error: Unsupported target for indexed assignment ("Optional[Dict[str, DelegatedRole]]")  [index]`

lukpueh commented on tests/repository_simulator.py:
```diff
@@ -354,21 +355,60 @@ def add_delegation(
+        succinct_roles = SuccinctRoles([], 1, bit_length, name_prefix)
+        delegator.delegations = Delegations({}, None, succinct_roles)
+
+        # Add target metadata for all bins.
```
Nit: 
```suggestion
        # Add targets metadata for all bins.
```

lukpueh commented on tests/repository_simulator.py:
```diff
@@ -354,21 +355,60 @@ def add_delegation(
         if role.name not in self.md_delegates:
             self.md_delegates[role.name] = Metadata(targets, {})
 
+    def add_succinct_roles_to_delegations(
```
You usually advocate for short names... I think  `add_succinct_roles` would be good enough.

MVrachev replied:
Yep, I do advocate for shorter names.
You are correct, that would work.

lukpueh commented on tests/repository_simulator.py:
```diff
@@ -354,21 +355,60 @@ def add_delegation(
+        if delegator_name == Targets.type:
+            delegator = self.targets
+        else:
+            delegator = self.md_delegates[delegator_name].signed
```
Maybe worth adding a helper for this if/else to be used here, in `add_target` and `add_delegation`.

lukpueh commented on tests/repository_simulator.py:
```diff
@@ -354,21 +355,60 @@ def add_delegation(
+    def add_succinct_roles_to_delegations(
+        self, delegator_name: str, bit_length: int, name_prefix: str
+    ) -> None:
+        """Add succinct roles info to a delegator with name "delegator_name"."""
```
I'd probably add a note that you are also bootstrapping empty bins metadata in this function.

lukpueh commented on tests/test_metadata_serialization.py:
```diff
@@ -174,6 +175,7 @@ def test_invalid_key_serialization(self, test_case_data: str) -> None:
     invalid_roles: utils.DataSet = {
         "no threshold": '{"keyids": ["keyid"]}',
         "no keyids": '{"threshold": 3}',
+        "wrong keyids type": '{"keyids": 1, "threshold": 3}',
```
This seems unrelated, but I guess it's okay to include it in this PR.

MVrachev replied:
Oh, I just noticed we are missing it and added it.
Will add this change to the commit with changes related to `DelegatedRole`.

MVrachev commented:
> Any thoughts about also changing `Root.{add, remove}_key`?

Probably it's best if we rename them as well, right?
As they are logically the same - adding or removing a key of a delegated role.
Any thoughts @jku on the above discussion and `Root.*_key()` API renames?

PS: Before any new updates on this pr I will wait for the full tests review and decision on what to do with the `Root` and `Targets` key API methods.

jku commented:
I think all *_key() methods should have the same logic (so yes root methods should change argument order and name if the targets methods do, even if they don't practically need to). 

Renaming to revoke_key() is acceptable I suppose? seems like best idea so far.

lukpueh commented on tests/test_updater_delegation_graphs.py:
```diff
@@ -425,6 +425,90 @@ def test_hash_bins_graph_traversal(
+            self.assertIsNone(targetfile)
+
+            # Check that the delegated roles were visited in the expected
+            # order and the corresponding metadata files were persisted
```
```suggestion
            # order and the corresponding metadata files were persisted
```

lukpueh replied:
Oops wrong line.

lukpueh replied:
Please ignore this one.

MVrachev commented:
Okay, when @lukpueh signals me that he has reviewed all of the tests and after merging #2017 I will update the pr addressing all comments and renaming the `{Root/Targets}*_key` methods.

lukpueh commented on tests/test_updater_delegation_graphs.py:
```diff
@@ -425,6 +425,90 @@ def test_hash_bins_graph_traversal(
+            target_path="bar",
+            expected_visited_bin="delegated-fc",
+        ),
+    }
```
Very cool test! I must admit though that it took me a while to understand what it's doing. I have some ideas to make this a bit easier. Feel free to adopt/ignore as you please.

- `expected_visited_bin` implied (to me) that it is some node along the tree traversal, but the tree is actually pretty shallow (root + leaves). Maybe we can just call it  `expected_bin`, as it is the bin we expect the target to fall into.
- Using `delegated-` as name prefix makes sense, but if you call it `bin-` it would further reduce complexity (e.g.: `expected_bin="bin-0"`) 
- Some `target_path` values ("missingpath", "othermissingpath") convey an irrelevant meaning, others don't ("foo", "doo", ...). This is confusing. Maybe you can use all placeholder names?
- A few lines of comments would make my suggestions from above less important. E.g. how `bit_length` determines both the total number of bins and the bin a `target_path` falls into, based on its hash prefix...
- Similarly/alternatively, some of this information could be encoded in the test case names, which are currently not very descriptive, e.g. instead of "case 6" it could say "bin count: 4, target bin index: 3"

lukpueh reviewed (changes requested):
Part 2/2 of tests review. Includes
```
tests/repository_simulator.py
tests/test_metadata_serialization.py
tests/test_updater_delegation_graphs.py
```

lukpueh commented on tests/test_updater_delegation_graphs.py:
```diff
@@ -425,6 +425,90 @@ def test_hash_bins_graph_traversal(
+            self._assert_files_exist(TOP_LEVEL_ROLE_NAMES)
+
+            # Looking for a non-existing targetpath forces updater
+            # to visit a correspondning delegated role
```
That's the one with the typo:
```suggestion
            # to visit a corresponding delegated role
```

MVrachev commented:
I tried addressing all comments made by @lukpueh.
Hope it's better now :)).
Ready for review.

jku commented:
> I tried addressing all comments made by @lukpueh.

I'm not against force-pushing large PRs, that's often still the best path forward... but if that is chosen (vs adding commits and potentially squashing after review), a clear list of changes since the last review (especially for the API changes) would be nice.

This is a bit of a review pain point in github as it's really hard to see changes between PR versions. I _think_ in this case it's this https://github.com/theupdateframework/python-tuf/compare/a1fb89c..c04ec75

MVrachev commented:
Yes, you are right that probably some summary of my changes will be good.
All of them were addressing Lukas comments, so those comments can give you some idea what I changed + I renamed the Root/Targets.remove_key() to Root/Targets.revoke_key() as we had discussed.

What I did ca be summarized like this:
1. Removed the `zero_padding` test in favor of a more complicated `test_get_roles_in_succinct_roles` test.
2. Moved a comment about bin names inside `test_is_delegated_role_in_succinct_roles` a few lines below.
3. Updated the test about `__eq__` implementation for SuccinctRoles after the merge of #2017.
4. Some typo fixes.
5. Renamed `RepositorySimulator. add_succinct_roles_to_delegations()` to `RepositorySimulator.add_succinct_roles()` and added a better docstring.
6. Added a helper `RepositorySimulator._get_delegator()` which removes code dublication in `add_target`, `add_delegation` and `add_succinct_roles` in `RepositorySimulator`.
7. Updated `test_succinct_roles_graph_traversal` with many suggestions by @lukas from this comment https://github.com/theupdateframework/python-tuf/pull/2010#discussion_r892305989
8. Renamed `Root/Targets.remove_key()` to `Root/Targets.revoke_key()` plus updated the pr description and commit description mentioning those API changes.

jku commented on tuf/api/metadata.py:
```diff
@@ -954,8 +961,8 @@ def add_key(self, role: str, key: Key) -> None:
 
-    def remove_key(self, role: str, keyid: str) -> None:
-        """Removes key from ``role`` and updates the key store.
+    def revoke_key(self, role: str, keyid: str) -> None:
```
this (and add_key()) are now named exactly like the Targets methods, but have exactly opposite argument order...

MVrachev replied:
Yes, and I was wondering should I change them just for the sake of consistency?
We changed the order just because we want to make `role` optional, but don't want to do that in `Root`.
So, what do you think?

lukpueh replied:
> Yes, and I was wondering should I change them just for the sake of consistency?

Yes, that seems to be what we agreed on [above](https://github.com/theupdateframework/python-tuf/pull/2010#issuecomment-1149807399).

jku replied:
Me earlier:
> I think all *_key() methods should have the same logic (so yes root methods should change argument order and name if the targets methods do, even if they don't practically need to).

I think it's bad if we have methods that look very similar but are not. So IMO either we don't change any argument order or we change all of them.

MVrachev replied:
Okay, will make the arguments in the same order.

MVrachev commented:
I have updated the pr by making `Root.*_key()` API methods with consistent naming and argument order as the ones in `Targets.*_key()`.

jku commented on tuf/api/metadata.py:
```diff
@@ -1285,6 +1297,8 @@ class DelegatedRole(Role):
         ``paths`` and ``path_hash_prefixes`` are mutually exclusive: both cannot be
         set, at least one of them must be set.
 
+    Exactly one of ``paths`` and ``path_hash_prefixes`` must be set.
```
nitpick: I think this is not needed now: the previous sentence seems to say the same thing.

jku commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1449,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
 
 
+class SuccinctRoles(Role):
+    """Succinctly defines a hash bin delegation graph.
+
+    A ``SuccinctRoles`` object describes a delegation graph that covers all
+    targets, distributing them uniformly over the delegated roles (i.e. bins)
+    in the graph.
+
+    The total number of bins is 2 to the power of the passed ``bit_length``.
+    Targets are assigned to bins by casting the left-most ``bit_length`` of
+    bits of the file path hash digest to int, using it as bin index between 0
+    and ``2**bit_length - 1``.
+
+    Bin names are the concatenation of the passed ``name_prefix`` and a
+    zero-padded hex representation of the bin index separated by a hyphen.
+
+    The passed ``keyids`` and ``threshold`` is used for each bin, and each bin
+    is 'terminating'.
+
+    For details: https://github.com/theupdateframework/taps/blob/master/tap15.md
```
I think there's still a lot to improve in this documentation: it contains loads of detail (like _"casting the left-most bit_length of bits of the file path hash digest to int, using it as bin index between 0 and 2**bit_length - 1"_), but probably does not help the user that much.

That said, I don't think we need to polish it in this PR, and should look at this more holistically: maybe the doc improvements should happen elsewhere.

lukpueh replied:
Haha, yes, I take the blame for this one, and I absolutely agree that it should become more readable. But I also think that the user of this interface needs to know how``bit_length`` functions, in order to choose an appropriate value.

Ideally the user isn't bothered with these details at all and just passes the expected amount of target files, and the tool figures out an appropriate ``bit_length``.

MVrachev replied:
Honestly, it was hard for me to describe what `SuccinctRoles` is without mentioning `bit_length` definition and what it represents. 
I feel like a short summary of how a target is assigned to a bin should be here or `get_role_for_target`. 
Now when I think about this... I think it does make more sense to be inside `get_role_for_target` as that's the actual API doing that.
Here it should be a more general description of `SuccinctRoles` and not so much how it does its magic.

jku commented on tuf/api/metadata.py:
```diff
@@ -1497,25 +1672,58 @@ def from_dict(cls, delegations_dict: Dict[str, Any]) -> "Delegations":
+        if self.roles is not None:
+            roles = [role_obj.to_dict() for role_obj in self.roles.values()]
+            res_dict["roles"] = roles
+        elif self.succinct_roles is not None:
```
This is about an edge case but... should this be just `else:`?

because I think we might prefer an exception at serialization if both roles and succinct_roles are None: that should not have happened.

MVrachev replied:
I agree with you, but `mypy` wants to know that `succinct_roles is not None` in order to call `succinct_roles.to_dict()` otherwise it complains.
It's absolutley the same situation in `DelegatedRole.to_dict()`.
Do you want me to add an additional `else` and throw an error if they are missing `paths/path_hash_prefixes` or `roles/succinct_roles` respectivly? 
I don't think it's necessary given we have a check in `__init__`.

jku commented on tuf/api/metadata.py:
```diff
@@ -1765,42 +1973,73 @@ def to_dict(self) -> Dict[str, Any]:
-        if self.delegations is None or role not in self.delegations.roles:
+        # Verify that our users are not using the old argument order.
+        if isinstance(role, Key):
+            raise ValueError("Argument order has changed")
```
nitpick: we shouldn't make guesses about how the error happened, we should just report what the situation is

jku reviewed (approved):
I think I'm happy with this -- left some nitpicks but they're not critical. The main question I have is documentation:
* The documentation currently does not make it clear that this feature is not a TUF specification feature: maybe it should?

As TODO items (that might exist already):
* we should improve the succinct delegation docs and update the example script -- this is something we should do before we release this
* for completeness, actual download tests should contain succinct delegations: it may make sense to refactor the download tests at that point

MVrachev commented:
Updated the pr with the following changes:
- added a note that `SuccinctRoles` is not supported by the spec
- addressed all other changes mentioned by @jku

jku commented on tuf/api/metadata.py:
```diff
@@ -1447,28 +1603,40 @@ class Delegations:
 
+    Exactly one of ``roles`` and ``succinct_roles`` must be set.
+
+    Note: succinct_roles is a feature not yet part of the specification.
```
API documentation should be direct: why are we telling this to the user of this API, and what can they do about it? Something like 

```suggestion
    Note that succinct roles is not a TUF specification feature yet and setting
    `succinct_roles` to a value makes the resulting metadata non-compliant.
    The metadata will not be accepted as valid by specification compliant 
    clients such as those built with python-tuf <= 1.1.0.
```

Could even move this to the actual argument description :shrug:

MVrachev commented:
Accepted @jku suggested how to document that `succinct_roles` is not in the spec yet and moved that documentation next to the `succinct_roles` argument description in `Delegations`.

jku commented:
Yes, looks good to me. Thanks everyone! The rough plan here is:
* Merge this PR: succinct delegations are then usable on "develop" branch
* Get any remaining polish done on TAP-15 and move it to "accepted" state https://github.com/theupdateframework/taps/pull/148
* release python-tuf with TAP-15 support (likely means v2.0.0)
* Maybe TAP-15 gets integrated into spec at some point?


@lukpueh, would you like to do a final review of the PR as the system thinks your review is pending? Checking  _"merge without waiting for requirements to be met"_ makes me feel criminal...

@MVrachev would you mind having a last look to identify any issues we may need to file for future work? 
* Updating the example code is one (if it does not exist already)
* Maybe also one for better parameterised target download tests that use succinct bins and normal bins -- there is a PR for the latter but I think we can do better

MVrachev commented:
> @MVrachev would you mind having a last look to identify any issues we may need to file for future work?
> 
> * Updating the example code is one (if it does not exist already)
> * Maybe also one for better parameterised target download tests that use succinct bins and normal bins -- there is a PR for the latter but I think we can do better

We have an issue https://github.com/theupdateframework/python-tuf/issues/1909 where we keep track of that stuff.
Only three items are left on the list and I am not sure there is a need for separate issues.

lukpueh commented on tests/test_updater_delegation_graphs.py:
```diff
@@ -425,6 +425,91 @@ def test_hash_bins_graph_traversal(
+        target_path: str
+        expected_target_bin: str
+
+    # By setting the bit_length the total number of is 2^bit_length.
```
```suggestion
    # By setting the bit_length the total number of bins is 2^bit_length.
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1435,6 +1447,150 @@ def is_delegated_path(self, target_filepath: str) -> bool:
+
+    def is_delegated_role(self, role_name: str) -> bool:
+        """Determines whether the given ``role_name`` is in one of
+        the delegated roles that ``SuccinctRoles`` represent.
```
```suggestion
        the delegated roles that ``SuccinctRoles`` represents.
```

lukpueh reviewed (approved):
LGTM! 💯
